### PR TITLE
Peer logic

### PR DESCRIPTION
**Pull request by CarolineZhou**

### CODE REVIEW

File: code-collab-app/src/editorpage.js [hunk 1]
```diff
@@ -6,39 +6,22 @@ import logo from './codeCollabLogo.png';
...
 class Editorpage extends React.Component {
   constructor(props){
     super(props);
 
     this.state = {
-      cellCount: 0,
-      cellText: (this.props.location.data) ? this.props.location.data : [],
-      cellLocked: [],
-      cellCurrentlyEditing: null,
+      cellCount: this.props.peer_data.cell_count,
```
Comment (CarolineZhou) [inline code-collab-app/src/editorpage.js:16]: We need these data on the peer_data side so i just injected peer_data store in this file and using them as prop values. The other way was causing some auto rendering issues, and this seems to be working fine.

File: code-collab-app/src/editorpage.js [hunk 2]
```diff
@@ -6,39 +6,22 @@ import logo from './codeCollabLogo.png';
...
   }
 
   componentDidMount() {
-    if(this.state.cellText.length === 0){
-      this.addEditorCell('print("Hello World!")');
-    }
-    this.props.location.peer.listen_for_req();
-  }
-
-  setCellText = (keyvalue, cellContents) => {
```
Comment (CarolineZhou) [inline code-collab-app/src/editorpage.js]: Replaced this with a method on the peer_data side, and now all the changes are made on the peer_data side and those changes trickle back to editor page's state by props.

File: code-collab-app/src/editorpage.js [hunk 3]
```diff
@@ -47,36 +30,22 @@ class Editorpage extends React.Component {
     if (!cellContents) {
       cellContents = '';
     }
-
-    this.setState({
```
Comment (CarolineZhou) [inline code-collab-app/src/editorpage.js:36]: Same as comment on line 29. Now we set the state in peer_data which comes back to this page by props.

File: code-collab-app/src/editorpage.js [hunk 4]
```diff
@@ -47,36 +30,22 @@ class Editorpage extends React.Component {
...
-    temp[key] = true;
-    this.setState({ cellLocked:temp });
+    this.props.peer_data.set_cell_locked(key, true);
   }
 
-  // Update cell text, then Unlock a cell so the user may edit it
+  // Unlock a cell so the user may edit it
   // key: the index of the cell to unlock
-  // updatedCellContents: updated cell text to apply to the cell
-  unlockEditorCell = (key, updatedCellContents) => {
```
Comment (CarolineZhou) [inline code-collab-app/src/editorpage.js]: Here we don't need update cell content because peer_data now have access to everything.

File: code-collab-app/src/editorpage.js [hunk 5]
```diff
@@ -205,4 +173,4 @@ class Editorpage extends React.Component {
   }
 }
 
-export default Editorpage;
+export default inject('peer_data')(observer(Editorpage));
```
Comment (CarolineZhou) [inline code-collab-app/src/editorpage.js:176]: injecting peer_data store directly here.

File: code-collab-app/src/homepage.js [hunk 6]
```diff
@@ -13,35 +13,16 @@ class Homepage extends React.Component {
   }
 
   handleJoinSession(session_info) {
+    this.props.peer_data.reset();
     this.props.peer_data.join_session(session_info.document_name, session_info.id);
-    // If only peer in the session, get doc data from tracker
-    if(session_info.peers.length === 0) {
```
Comment (CarolineZhou) [inline code-collab-app/src/homepage.js:18]: This logic is now moved into peer_data.

File: code-collab-app/src/homepage.js [hunk 7]
```diff
@@ -13,35 +13,16 @@ class Homepage extends React.Component {
...
-      this.props.peer_data.get_document_data(session_info.id);
-      this.props.peer_data.socket.on('rcv_doc_data', (data) => {
-        this.props.peer_data.doc_data = data.doc;
-        this.props.history.push({
-          pathname: '/editor',
-          data: this.props.peer_data.doc_data,
```
Comment (CarolineZhou) [inline code-collab-app/src/homepage.js:24]: Not passing any data here because peer_data is directly injected into editor component now.

File: code-collab-app/src/homepage.js [hunk 8]
```diff
@@ -13,35 +13,16 @@ class Homepage extends React.Component {
...
-        pathname: '/editor',
-        data: this.props.peer_data.doc_data,
-        peer: this.props.peer_data
-      });
-    }
+    this.props.history.push({pathname: '/editor'});
   }
 
   handleNewSession() {
+    this.props.peer_data.reset();
```
Comment (CarolineZhou) [inline code-collab-app/src/homepage.js:22]: This resets all the existing data and closes all older peer connections. And tracker connection is kept as it is.

File: tracker/index.js [hunk 9]
```diff
@@ -53,9 +53,9 @@ io.on("connection", peer => {
   });
 
   // Get document data request
-  peer.on('get_doc_data', session_id => {
```
Comment (CarolineZhou) [inline tracker/index.js:56]: Changed this to match peer to peer request and response format.

File: code-collab-app/src/editorpage.js [hunk 10]
```diff
@@ -32,70 +30,68 @@ class Editorpage extends React.Component {
...
   }
 
-  // Unlocks a cell so the user may edit it
+  // Unlock a cell so the user may edit it
   // key: the index of the cell to unlock
   unlockEditorCell = (key) => {
-    let temp = this.state.cellLocked;
-    temp[key] = false;
-    this.setState({ cellLocked:temp });
+    this.props.peer_data.set_cell_locked(key, false);
```
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:47]: I added this in my PR (https://github.com/paulchenpmc/codeCollab/pull/18/files), but before we unlock this we should update the cell contents. So that means this function should supply the updatedCellContents as an argument when called from the peer logic.

Also I'm getting confused looking at this PR. I see some elements from my PR, but it's also missing some? For example in this function in my PR there is a call to `this.setCellText(key, updatedCellContents);` but that is missing here. However, the `onBlur` event from my PR is in this.

Another example, this PR is entirely missing the `setCellText` function. I have some serious concerns about how you merged these two PRs together,

File: code-collab-app/src/editorpage.js [hunk 11]
```diff
@@ -32,70 +30,68 @@ class Editorpage extends React.Component {
...
+
   }
 
   // Requests lock on cell from all peers in session
   // key: index of cell
   requestEditorCellLock = (key) => {
     // TODO - implement socket.io request to lock this cell on all other peers in session
+    // Will probably involve calling lockEditorCell()
     console.log('Requesting cell lock from all peers for cell ' + key);
```
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:56]: Still todo on the peer logic side I presume? This is one of the events that will fire when a cell is selected, so it will need to call some peer logic to obtain a consensus lock from all peers before editing may occur.

File: code-collab-app/src/editorpage.js [hunk 12]
```diff
@@ -32,70 +30,68 @@ class Editorpage extends React.Component {
...
     return true;
   }
 
   // Broadcast cell update and release lock for all other peers in session
   // key: index of cell
   // cellContents: text contents of the cell update
-  broadcastCellUpdate = (key, cellContents) => {
+  broadcastCellUpdate = (key) => {
     // TODO - implement socket.io broadcast to update cell for all other peers and release lock
+    // Will probably involve calling unlockEditorCell()
```
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:65]: Again, an event that fires when the user leaves a cell (I will implement timeout soon to trigger this as well). Needs to call some peer side logic to send this update to all peers in session.

File: code-collab-app/src/editorpage.js [hunk 13]
```diff
@@ -138,10 +130,11 @@ class Editorpage extends React.Component {
         id="outlined-multiline-flexible"
         label={cellLabel}
         multiline
-        onFocus={(e) => this.selectCellEvent(e, keyvalue)}
+        onFocus={(e) => this.selectCellEvent(e, keyvalue)} // When cursor enters cell
+        onBlur={(e) => this.leaveCellEvent(e, keyvalue)} // When cursor leaves cell
```
Comment (BKDuncan) [inline code-collab-app/src/editorpage.js:134]: Is this the `onBlur` you are looking for @paulchenpmc ?

Comment (BKDuncan) [reply code-collab-app/src/editorpage.js:134]: Nvm definitely misread your comment 😂

File: code-collab-app/src/editorpage.js [hunk 14]
```diff
@@ -47,36 +30,22 @@ class Editorpage extends React.Component {
...
+    this.props.peer_data.set_cell_locked(key, true);
   }
 
-  // Update cell text, then Unlock a cell so the user may edit it
+  // Unlock a cell so the user may edit it
   // key: the index of the cell to unlock
-  // updatedCellContents: updated cell text to apply to the cell
-  unlockEditorCell = (key, updatedCellContents) => {
-    // Update cell with new text
-    this.setCellText(key, updatedCellContents);
```
Comment (BKDuncan) [inline code-collab-app/src/editorpage.js]: @paulchenpmc if you look at the last commit instead of the change history you can see it was removed

File: code-collab-app/src/editorpage.js (hunk 11 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:56]: yes, this is still to come.

File: code-collab-app/src/editorpage.js (hunk 12 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:65]: yes, this is still to come.

File: code-collab-app/src/editorpage.js (hunk 10 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:47]: I merged your PR first and then changed the logic so that everything is available in peer_data, which is why you see some bits but not others. As Bailey mentioned in the comment below, you can look at the last commit in this PR to get a better idea.

The reason why we don't need `updatedCellContents` in `setCellText` is because the event listener from `Textfield` now directly calls a method in peer_data everytime anything in the cell is changed. So we should already have up to date cell info in peer_data before we leave the cell.

Again, the setCellText was removed because the `TextField` now directly calls a function in peer_data.

To summarize, Editor page is now have an injected copy of peer_data store. Any changes that are made in the Editor page flow towards peer_data store and then trickle back to Editor page state via props.

Sorry, I had to do all of this because we need all the updated data in peer_data so that it can be communicated with other peers.

Again, you can look at the last commit in this peer, and ask me if anything doesn't make sense.

File: code-collab-app/src/storage/peer_data.js [hunk 15]
```diff
@@ -68,71 +109,150 @@ const peer_data = observable({
...
+        this.session_peers_conn = this.session_peers_conn.filter(conn => conn.peer !== data_connection.peer);
+    },
+
+    upload_document(new_data) {
+        this.doc_data.push(...new_data);
     },
 
     // Get document data from the tracker if no peer active for the document
     get_document_data(session_id) {
         this.tracker.emit('get_doc_data', session_id);
```
Comment (chantalmlchung) [inline code-collab-app/src/storage/peer_data.js:220]: If tracker is no longer handling an event called 'get_doc_data', we probably don't need this?

Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js:220]: That's right, I will remove it in the next set of changes.

File: code-collab-app/src/homepage.js (hunk 7 above)
Comment (chantalmlchung) [reply code-collab-app/src/homepage.js:24]: Ok sounds good. But just one thing, since these socket events are asynchronous, have we tested to make sure the doc_data that the editor page is receiving is from the tracker and not the default (empty array)? I had this here to make sure we wait for the tracker's data before we send data to the editor page so I just want to make sure this is still being handled.

Comment (CarolineZhou) [reply code-collab-app/src/homepage.js:24]: It is receiving the doc_data from the peer_data store as a prop, so any changes made by sockets will update that and it will be pushed to Editor page. So I believe it is all good, I will test some more to make sure.

File: code-collab-app/src/storage/peer_data.js [hunk 16]
```diff
@@ -68,71 +109,150 @@ const peer_data = observable({
...
-                this.session_peers = list;
-                this.request_doc_data();
-            });
         });
     },
 
-    //  request for current session's doc data 
-    //  from a fellow peer
+    //  request doc data from a peer
     request_doc_data() {
```
Comment (chantalmlchung) [inline code-collab-app/src/storage/peer_data.js:118]: I might be missing something here, but how does the peer know which document to request if we don't provide the document name, etc?

File: code-collab-app/src/homepage.js (hunk 7 above)
Comment (chantalmlchung) [reply code-collab-app/src/homepage.js:24]: How are we guaranteeing that the editor page will have the correct document data when we load it since we aren't passing it any data? (i.e., when does the peer request the document data from the other peers or tracker? before or after loading the page? how does the peer know what document to retrieve?)

File: code-collab-app/src/storage/peer_data.js (hunk 16 above)
Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js:118]: When we have joined a session, all the peers only have access to one document. So we don't really need to know which document to get because there is only one in each session.

This is called after we have joined a session, and aimed towards one of the peers in the session.

Comment (chantalmlchung) [reply code-collab-app/src/storage/peer_data.js:118]: Ok so when does a peer join a session? Is this after the end user clicks on one of the "Active Sessions" on the homepage? If so, how do we guarantee the editor page will have the correct data?

File: code-collab-app/src/homepage.js (hunk 7 above)
Comment (CarolineZhou) [reply code-collab-app/src/homepage.js:24]: Each session has one document, and we do know which session we are joining. In case of asking the peers, all the peers in the session will have the same document so we will get the right one. And in case of the tracker, we do provide the session id (peer_data has access to session id).

File: code-collab-app/src/storage/peer_data.js (hunk 16 above)
Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js:118]: Yes, when the user clicks on one of the existing sessions, we join that session. We get a list of peers from the tracker for that session and then ask for data from the first peer. If no peers exists in the list that was returned by the tracker, then we ask the tracker for the document data using the session id.

File: code-collab-app/src/editorpage.js [hunk 17]
```diff
@@ -32,70 +30,68 @@ class Editorpage extends React.Component {
...
-    // Save the cell currently being edited
-    this.setState({ cellCurrentlyEditing:key });
+    // Remember the cell currently being edited
+    this.props.peer_data.set_current_cell(key);
+  }
+
+  // Event handler for cursor leaving a cell. Broadcasts update for that cell and releases lock.
+  // e: onBlur event
+  // key: index of cell
+  leaveCellEvent = (e, key) => {
```
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:84]: key is an unused arg here and can be removed

File: code-collab-app/src/storage/peer_data.js (hunk 16 above)
Comment (chantalmlchung) [reply code-collab-app/src/storage/peer_data.js:118]: Oh ok I see now. ! missed the "join_session" line from the homepage so I was wondering how the peer would know a user joined but it makes sense now that I saw that line.

Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js:118]: Perfect :)

File: code-collab-app/src/editorpage.js [hunk 18]
```diff
@@ -32,70 +30,68 @@ class Editorpage extends React.Component {
...
+    e.preventDefault();
+    // Get index of last cell edited
+    const lastCellEdited = this.state.cellCurrentlyEditing;
```
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:87]: When I tested this branch it seems the `cellCurrentlyEditing` and `cellCount` state variables never gets updated in the editorpage.

Testing indicates `this.props.peer_data.set_current_cell(key);` is called, but maybe this doesn't make it down to editorpage somehow?

File: code-collab-app/src/storage/peer_data.js [hunk 19]
```diff
@@ -68,71 +109,150 @@ const peer_data = observable({
...
+    update_existing_cell(index, value){
+        this.doc_data[index] = value;
+    },
+
+    set_current_cell(index){
+        this.current_cell = index;
+    },
+
+    increment_cell_count(){
+        this.count += 1;
```
Comment (paulchenpmc) [inline code-collab-app/src/storage/peer_data.js:238]: Pretty sure this is typo?

```suggestion
        this.cell_count += 1;
```

File: code-collab-app/src/editorpage.js (hunk 10 above)
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:47]: Typo? This is how it is spelled in peer_data.js

Also we need to update the cell contents before we release the lock, but based on your response it sounds like we will end up doing this on the peer_data side.

```suggestion
    this.props.peer_data.set_cell_lock(key, false);
```

File: code-collab-app/src/editorpage.js [hunk 20]
```diff
@@ -32,70 +30,68 @@ class Editorpage extends React.Component {
...
   }
 
   // Locks a cell so the user cannot edit it
   // key: the index of the cell to lock
   lockEditorCell = (key) => {
-    let temp = this.state.cellLocked;
-    temp[key] = true;
-    this.setState({ cellLocked:temp });
+    // Lock cell so user may not edit
+    this.props.peer_data.set_cell_locked(key, true);
```
Comment (paulchenpmc) [inline code-collab-app/src/editorpage.js:41]: Same typo as below
```suggestion
    this.props.peer_data.set_cell_lock(key, true);
```

File: code-collab-app/src/editorpage.js (hunk 10 above)
Comment (paulchenpmc) [reply code-collab-app/src/editorpage.js:47]: Makes sense. I will resolve this thread.

File: code-collab-app/src/editorpage.js (hunk 14 above)
Comment (paulchenpmc) [reply code-collab-app/src/editorpage.js]: Roger that

File: code-collab-app/src/editorpage.js (hunk 18 above)
Comment (paulchenpmc) [reply code-collab-app/src/editorpage.js:87]: Further testing indicates `cellText` and `cellLocked` state variables seem to update fine and trickle down to the editorpage

File: code-collab-app/src/editorpage.js (hunk 17 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:84]: Okay, I will remove it.

File: code-collab-app/src/editorpage.js (hunk 18 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:87]: Yeah, I am not quite sure. I think it is because those variables not really used anywhere in rendering. So to get around this issue, I am exposing methods to get the state of `cellCurrentlyEditing` and `cellCount` directly from the peer_data store, and remove them as state variables in Editor page. Thoughts?

File: code-collab-app/src/storage/peer_data.js (hunk 19 above)
Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js:238]: I will correct it in the next commit, thanks!

File: code-collab-app/src/editorpage.js (hunk 10 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:47]: Yup, this will be corrected in the next commit as well. Thank you!

File: code-collab-app/src/editorpage.js (hunk 20 above)
Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:41]: Yup, this will be corrected in the next commit as well. Thank you!

File: code-collab-app/src/editorpage.js (hunk 18 above)
Comment (paulchenpmc) [reply code-collab-app/src/editorpage.js:87]: @BKDuncan this might be right up your alley. Thoughts on best practice? Kinda seems like the editorpage state vars aren't properly referencing the peer_data ones.

Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:87]: Also, I don't seem to find the usage of cellCount anywhere. So I am going to remove it from the editor page for now and we can add a method in the peer_data store if we need it later.

Comment (paulchenpmc) [reply code-collab-app/src/editorpage.js:87]: Good catch, I think `cellCount` was a remnant from a previous architecture of rendering the cells.

Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js]: This is fixed in the latest commit.

File: code-collab-app/src/storage/peer_data.js (hunk 15 above)
Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js]: Removed in the lastest commit.

**Comment by CarolineZhou** [comment]
The last two commits fix some bugs and also sets up updates to other peers when the cell is changed or a new cell is added.

Left is the lock/unlock logic so that only one peer can edit a cell at any time.

File: code-collab-app/src/editorpage.js [hunk 21]
```diff
@@ -6,24 +6,20 @@ import logo from './codeCollabLogo.png';
...
   constructor(props){
     super(props);
 
     this.state = {
-      cellCount: 0,
-      cellText: (this.props.location.data) ? this.props.location.data : [],
-      cellLocked: [],
-      cellCurrentlyEditing: null,
+      cellText: this.props.peer_data.doc_data,
+      cellLocked: this.props.peer_data.cell_locked,
```
Comment (BKDuncan) [inline code-collab-app/src/editorpage.js:17]: @paulchenpmc @CarolineZhou Turning a `prop` into a `state` works okay as long as we never use `setState`, but just a style suggestion. Since we already have these variables in `props`, we don't really need these variables to be turned into `state` variables.

In that case we could:
- Delete the whole constructor method
- Replace all references to `this.state.cellText` with `this.props.peer_data.doc_data`
- Replace all references to `this.state.cellLocked` with `this.props.peer_data.cell_locked`

Comment (CarolineZhou) [reply code-collab-app/src/editorpage.js:17]: Sure, I will change that in the next set of commits. This PR already have some conflicts so I want to finish merging this first. After that I will make those changes along with the next functionality I am working on.

**Comment by CarolineZhou** [comment]
Weirdly enough the conflicts were just the normal changes I made, and Git somehow got confused. Anyways, it is all fixed now. So merging this to master.

File: code-collab-app/src/storage/peer_data.js [hunk 22]
```diff
@@ -68,71 +110,172 @@ const peer_data = observable({
...
+                });
+            });
+        }
     },
 
-    // Get document data from the tracker if no peer active for the document
-    get_document_data(session_id) {
-        this.tracker.emit('get_doc_data', session_id);
+    get_current_cell(){
+        return this.current_cell;
```
Comment (BKDuncan) [inline code-collab-app/src/storage/peer_data.js:264]: Just FYI. All the variables in `peer_data` store are accessible in the react component once its injected, so we don't really need getter/setter functions unless its useful to hide a bunch of calls to other peers.

E.G. 
- To get current cell in editor page we could use: `this.props.peer_data.current_cell`
- To change current cell in editor page we could use: `this.props.peer_data.current_cell = 1`

Comment (CarolineZhou) [reply code-collab-app/src/storage/peer_data.js:264]: Yeah I might remove all of this later on. I kept it in because we were using the local states. But I think I will remove most of them with an exception of the few where a lot of stuff is changed, that I will leave in the peer_data store to keep the mess out of editor page.
